### PR TITLE
feat: add NVIDIA NIM ranker support

### PR DESCRIPTION
**Pull request by mattf**
### Proposed Changes:

NVIDIA NIMs (hosted on https://build.nvidia.com and downloadable) support reranking models. This adds an NvidiaRanker component to access those models.

### How did you test it?

- [x] unit tests
- [x] integration tests of hosted NIM
- [x] integration tests of downloaded NIM

### Notes for the reviewer

1. should the input documents be mutated to add a score?
2. the ranker endpoints are not standardized, so i introduced an endpoint arg to rank(). an alternative is to refactor all components to pass the full inference endpoint to the backend.

### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.

### CODE REVIEW

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (davidsbatista) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:239]:
```suggestion
        )
```

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (davidsbatista) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:241]: Let's rather add it here for better readability

```suggestion
        for item in sorted_indexes_and_scores[:top_k]:
```

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (davidsbatista) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:242]: Since we are only updating the `doc.score` field I would say there's no need to copy

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (davidsbatista) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:224]: Why a `bool` ?

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (davidsbatista) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:9]: Let's remove this commented spec

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (davidsbatista) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:127]: we dropped the use of plain text API, keys, you can simply to this, and the user sets it with an environment variable  or pass with this parameter

```suggestion
        self._api_key = Secret.from_env_var("NVIDIA_API_KEY")
```

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (davidsbatista) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:140]: You can remove this and initialised like in my suggestions

**Review by davidsbatista — COMMENTED** [review]
Hi @mattf! Many thanks for your contribution, I've left a few comments

**Comment by davidsbatista** [comment]
@anakin87 can you also have a look on this PR?

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (mattf) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:224]: `isinstance(True/False, int)` -> `True`

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (mattf) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:127]: should i entirely remove the `api_key` param to `__init__`?

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (mattf) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:140]: there are two modes of operation -

 1. hosted - calls to api.nvidia.com, these require an api_key
 2. local - calls to some user known api_url, these do not require an api_key

i would like to only require the api_key in the first case. ok to keep this?

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (davidsbatista) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:137]:
```suggestion
            api_key=self._api_key.resolve_value(),
```

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (davidsbatista) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:94]: ah ok, I see, then here it's enough to to `api_key = Secret.from_env_var("NVIDIA_API_KEY")` and then call the `.resolve_value()` in the run() method

Comment (mattf) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:94]: is it sufficient that resolve_value is called in the NimBackend init?

see https://github.com/deepset-ai/haystack-core-integrations/pull/1023/files#diff-4f9b82acfb7e3f7ebe10dc8f69f7f3e26a01b24e34c7b8a74cfd55c372cda5aaR41

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (mattf) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:178]: if we allow True/False for top_k, that's what will happen.

if it's ok to accept True/False as 1/0, i can remove the bool portion of the condition.

imho, a user who pass in a bool is on the path to confusion. i'll defer to your judgement.

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (davidsbatista) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:94]: ok, it's called there, I missed that

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (silvanocerza) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:28]: This wrong, `api_key` must be a `Secret`. I would change this like so:

```suggestion
        api_key=Secret.from_token("your-api-key"),
```

You also need to add its import.

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (silvanocerza) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:46]: Why is this `Optional` if you set a default and enforce it to be set down below?

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (silvanocerza) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:144]:
```suggestion
    @component.output_types(documents=List[Document])
```

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (silvanocerza) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:178]: The `bool` check is not necessary, you're already checking if it's not an `int`. Also the `run` type hint clearly says that it mus be an `int`.

```
In [1]: top_k = 1

In [2]: not isinstance(top_k, int) or isinstance(top_k, bool)
Out[2]: False

In [3]: top_k = True

In [4]: not isinstance(top_k, int) or isinstance(top_k, bool)
Out[4]: True

In [5]: top_k = 0.1

In [6]: not isinstance(top_k, int) or isinstance(top_k, bool)
Out[6]: True

In [7]: top_k = "1"

In [8]: not isinstance(top_k, int) or isinstance(top_k, bool)
Out[8]: True
```

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (silvanocerza) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:75]: No need to check it's a `bool`, it must be an `int` in any case as the type hint above implies too.

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (silvanocerza) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:182]: Can be simplified.

```suggestion
        top_k = top_k or self._top_k
```

File: integrations/nvidia/src/haystack_integrations/utils/nvidia/nim_backend.py
Comment (silvanocerza) [inline integrations/nvidia/src/haystack_integrations/utils/nvidia/nim_backend.py:140]:
```suggestion
        url = endpoint or f"{self.api_url}/ranking"
```

File: integrations/nvidia/tests/test_ranker.py
Comment (silvanocerza) [inline integrations/nvidia/tests/test_ranker.py:1]: Would be great if we could name tests in a more descriptive way so we can easily understand what they're testing. `test_documents_typeerror2` is really not a great name for a test.

File: integrations/nvidia/tests/test_base_url.py
Comment (silvanocerza) [inline integrations/nvidia/tests/test_base_url.py:64]: What is this about?

**Review by silvanocerza — CHANGES_REQUESTED** [review]
Left some comments, there are some imperfectiosn to fix.

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (mattf) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:182]: my thinking is `run(..., top_k=...)` should always override `NvidiaRanker(..., top_k=...)` including when `run(..., top_k=0)`

`top_k or self._top_k` -> `self._top_k` when `top_k=0`

Comment (silvanocerza) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:182]: Ah right, didn't consider `top_k` could be `0`. 🤦 

Disregard this.

File: integrations/nvidia/tests/test_base_url.py
Comment (mattf) [reply integrations/nvidia/tests/test_base_url.py:64]: a note to myself, which i decided to skip because it was a refactor. i'll remove it.

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (mattf) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py]: i changed it to be None and set default within the constructor

File: integrations/nvidia/tests/test_ranker.py
Comment (mattf) [reply integrations/nvidia/tests/test_ranker.py:1]: i combined the two tests into one

**Comment by shadeMe** [comment]
@mattf FYI - we avoid force-pushing to a branch once a review's been published. Otherwise, the review comments lose their anchors and it's difficult for the reviewer to diff changes since their previous review.

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (silvanocerza) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:48]: I still don't get the benefit of this being `Optional[str]`, can't we just set it to `str` and a have proper default?

**Comment by mattf** [comment]
> @mattf FYI - we avoid force-pushing to a branch once a review's been published. Otherwise, the review comments lose their anchors and it's difficult for the reviewer to diff changes since their previous review.

understood. it was one of the options the github presented to catch up to main when the pr was out of date. it may be possible to disable this in the project settings.

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (mattf) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:48]: this is future proofing. in the case of a downloaded NIM, the default can be detected by querying /v1/models on the api_url.

an alternative is to `model: str = _DEFAULT_MODEL` and in the future check `api_url is not None and model == _DEFAULT_MODEL` before getting a new default based on `{api_url}/models`. a corner case being if the user actually specified a `model=...` that happened to match `_DEFAULT_MODEL`. my inclination is to keep it `None` so we can tell if the user passed `model` or not (sans the case where they pass `model=None`).

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (shadeMe) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:81]: Is this not something we can handle in this PR?

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (shadeMe) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:49]: Can we replace this with an enum like we've done in the [embedders](https://github.com/deepset-ai/haystack-core-integrations/blob/36c5401f2c8a3ae4465afcc00702e0d0f5cacdc7/integrations/nvidia/src/haystack_integrations/components/embedders/nvidia/truncate.py#L4)?

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (shadeMe) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:106]:
```suggestion
        :returns: A dictionary containing the ranker's attributes.
```

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (shadeMe) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:123]:
```suggestion
        :returns: The deserialized ranker.
```

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (shadeMe) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:145]: Let's exit early if `self._initialized` is already `True`.

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (shadeMe) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:165]:
```suggestion
        :returns: A dictionary containing the ranked documents.
```

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (shadeMe) [inline integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:183]:
```suggestion
        if not all(isinstance(doc, Document) for doc in documents):
            msg = "Ranker expects the `documents` parameter to be a list of Document objects."
            raise TypeError(msg)
        if top_k is not None and not isinstance(top_k, int):
            msg = "Ranker expects the `top_k` parameter to be an integer."
            raise TypeError(msg)

        if len(documents) == 0:
            return {"documents": []}            
```
To group the errors.

**Review by shadeMe — CHANGES_REQUESTED** [review]
Just a few small changes before we can merge it.

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (mattf) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py]: i added `RankerTruncateMode(str, Enum)`

i did not use `from_str`, instead relying on `Enum` constructor. let me know if i should implement `from_str`.

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (mattf) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:82]: i would prefer to do this in a follow up feat, if that's ok with you

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (mattf) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py]: updated to skip second warm_ups

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (shadeMe) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py]: We'd prefer to keep it consistent across the codebase (it's a common pattern that we use), so it'd be great if we could add `from_str`, use `truncate: Optional[Union[RankerTruncateMode, str]] = None` as the init param and call `RankerTruncateMode.from_str` if the passed value is a string. You can basically use the same code as the one in the embedder truncate enum.

File: integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py
Comment (shadeMe) [reply integrations/nvidia/src/haystack_integrations/components/rankers/nvidia/ranker.py:82]: Works for me.

**Review by shadeMe — APPROVED** [review]
LGMT! Thanks for the quick turnaround!

**Pull request merged**
